### PR TITLE
WebXR to OpenXR

### PR DESCRIPTION
**Pull request by syntheticmagus**

### CODE REVIEW

File: Library/Dependencies/xr/Include/XR.h [hunk 1]
```diff
@@ -25,6 +25,11 @@ namespace xr
     class System
     {
     public:
+        // These values are taken from Microsoft's OpenXR sample and are used in conjunction with the depth extension. Values are reversed, as in example.
+        // https://github.com/microsoft/OpenXR-SDK-VisualStudio/blob/42172676a9388f02ce5a776c89a62cecf95429bd/samples/BasicXrApp/OpenXrProgram.cpp#L226
```
Comment (syntheticmagus) [inline Library/Dependencies/xr/Include/XR.h]: These comments are no longer accurate; remove them.

File: Library/Source/NativeEngineImpl.h [hunk 2]
```diff
@@ -174,6 +207,7 @@ namespace babylon
         void Bind(FrameBufferData* data)
         {
             assert(m_boundFrameBuffer == nullptr);
+
```
Comment (syntheticmagus) [inline Library/Source/NativeEngineImpl.h]: nit: Revert line?

File: Library/Source/XrPlugin.cpp [hunk 3]
```diff
@@ -18,62 +20,146 @@ namespace
...
+        0.f, 0.f, 1.f, 0.f,
+        0.f, 0.f, 0.f, 1.f
+    };
+
+    // From https://github.com/microsoft/OpenXR-SDK-VisualStudio/blob/master/samples/XrUtility/XrMath.h#L332-L335
+    // 2 * n / (r - l)        0                     0                    0
+    // 0                      2 * n / (t - b)       0                    0
+    // (r + l) / (r - l)      (t + b) / (t - b)     f / (n - f)         -1
+    // 0                      0                     n*f / (n - f)        0
+    // Except we're row-major, so we transpose.
```
Comment (syntheticmagus) [inline Library/Source/XrPlugin.cpp]: This comment is neither accurate nor relevant; delete it.

File: TestApp/Scripts/experience.js [hunk 4]
```diff
@@ -10,14 +10,16 @@ function CreateBoxAsync() {
 }
 
 function CreateSpheresAsync() {
```
Comment (syntheticmagus) [inline TestApp/Scripts/experience.js]: Revert changes to this function.

Comment (syntheticmagus) [reply TestApp/Scripts/experience.js]: Also most of the rest of the changes to this file. Make the OpenXR thing an option similar to existing options.

File: Library/Dependencies/xr/Include/XR.h [hunk 5]
```diff
@@ -95,12 +98,13 @@ namespace xr
...
             std::unique_ptr<Frame> GetNextFrame(bool& shouldEndSession, bool& shouldRestartSession);
             void RequestEndSession();
+            std::pair<size_t, size_t> GetWidthAndHeightForViewIndex(size_t viewIndex) const;
 
         private:
             std::unique_ptr<Impl> m_impl{};
         };
 
-        System();
+        System(const char* = "OpenXR Experience");
```
[the remark this answers is not in the document]
Comment (syntheticmagus) [reply Library/Dependencies/xr/Include/XR.h]: It's a thing OpenXR provides. At present, the only place I've found it to manifest is in Task Manager, though I assume it's possible for runtimes to use it other ways. The spec isn't super clear about what platforms are supposed to do with this.

File: Library/Source/XrPlugin.cpp [hunk 6]
```diff
@@ -200,20 +280,789 @@ namespace babylon
...
+
+        private:
+            static inline Napi::FunctionReference constructor{};
+
+            // Lifetime control to prevent the cleanup of the NativeEngine while XR is still alive.
+            Napi::ObjectReference m_jsEngineReference{};
+
+            Napi::Value InitializeXRLayerAsync(const Napi::CallbackInfo& info)
+            {
+                // TODO BAD DESIGN: Having this here introduces something of a circular dependency with the session
```
Comment (bghgary) [inline Library/Source/XrPlugin.cpp]: do you want to fix this before checking in?

File: Library/Source/XrPlugin.cpp [hunk 7]
```diff
@@ -200,20 +280,789 @@ namespace babylon
...
+                // remove the circular dependency.
+                auto& session = *XRSession::Unwrap(info[0].As<Napi::Object>());
+                session.SetEngine(m_jsEngineReference.Value());
+
+                auto xrLayer = XRWebGLLayer::New();
+                session.InitializeXrLayer(xrLayer);
+                info.This().As<Napi::Object>().Set("xrLayer", xrLayer);
+
+                auto deferred = Napi::Promise::Deferred::New(info.Env());
+                deferred.Resolve(Napi::Value::From(info.Env(), 0)); // TODO: Why is it REQUIRED that I pass a valid value here?
```
Comment (bghgary) [inline Library/Source/XrPlugin.cpp]: You should probably resolve with `undefined`

File: Library/Source/RuntimeUWP.cpp [hunk 8]
```diff
@@ -18,11 +20,15 @@ namespace babylon
 
     /*RuntimeUWP::RuntimeUWP(ABI::Windows::UI::Xaml::Controls::ISwapChainPanel* panel, const std::string& rootUrl)
         : Runtime{ std::make_unique<RuntimeImpl>(from_abi<winrt::Windows::UI::Xaml::Controls::SwapChainPanel>(panel), rootUrl) }
-    {
-    }*/
+    {}*/
 
     void RuntimeImpl::ThreadProcedure()
     {
+        this->Execute([](RuntimeImpl& runtimeImpl)
```
[the remark this answers is not in the document]
Comment (syntheticmagus) [reply Library/Source/RuntimeUWP.cpp]: This queues up the work on the task chain to be run once the dispatcher starts getting drained, which will happen inside BaseThreadProcedure(). Essentially, this is just a non-invasive way to get the Env() and ensure this is only run after the JavaScript is done initializing.

File: Library/Source/XrPlugin.cpp [hunk 9]
```diff
@@ -18,62 +20,145 @@ namespace
...
+        0.f, 0.f, 1.f, 0.f,
+        0.f, 0.f, 0.f, 1.f
+    };
+
+    inline std::array<float, 16> createProjectionMatrix(const xr::System::Session::Frame::View& view)
+    {
+        const float n{ view.DepthNearZ };
+        const float f{ view.DepthFarZ };
+
+        const float r = std::tanf(view.FieldOfView.AngleRight) * n;
```
[the remark this answers is not in the document]
Comment (syntheticmagus) [reply Library/Source/XrPlugin.cpp]: Leftovers from copy-paste. Fixed in these two methods.

File: Library/Source/NativeEngine.cpp [hunk 10]
```diff
@@ -1067,6 +1073,14 @@ namespace babylon
         });
     }
 
+    void NativeEngine::Impl::Dispatch(std::function<void()> function)
+    {
+        m_runtimeImpl.Execute([function = std::move(function)](auto&)
```
[the remark this answers is not in the document]
Comment (syntheticmagus) [reply Library/Source/NativeEngine.cpp]: https://github.com/BabylonJS/BabylonNative/issues/63

File: Library/Source/XrPlugin.cpp [hunk 11]
```diff
@@ -18,62 +20,145 @@ namespace
...
+                callback(*m_frame);
+                EndFrame();
+            });
+        }
+
+        template<typename CallableT>
+        void Dispatch(CallableT&& callable)
+        {
+            m_engineImpl->Dispatch(callable);
+        }
```
[the remark this answers is not in the document]
Comment (syntheticmagus) [reply Library/Source/XrPlugin.cpp]: Made this match the NativeEngine version instead, as that version more closely resembles the existing methods in RuntimeImpl.

**Pull request merged**